### PR TITLE
Update documentation regarding `PEP 681` support with `mypy`

### PR DESCRIPTION
**Pull request by Viicos**
See https://mypy-lang.blogspot.com/2023/03/mypy-111-released.html
This also fixes a small typo

### Checklist

This pull request is:

- [x] A documentation / typographical error fix
	- Good to go, no issue or tests are needed
- [ ] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [ ] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

**Have a nice day!**

### CODE REVIEW

File: doc/build/orm/dataclasses.rst
Comment (CaselIT) [inline doc/build/orm/dataclasses.rst:40]: could probably update the date too

Comment (Viicos) [reply doc/build/orm/dataclasses.rst]: https://github.com/sqlalchemy/sqlalchemy/pull/9463/commits/3f0de536af822b2a78ad5cfb289728b277d92995, added `fully` as it seems other type checkers have basic support (e.g. pyre).

File: doc/build/orm/dataclasses.rst
Comment (zzzeek) [inline doc/build/orm/dataclasses.rst:41]: I would take the word "limited" out , at this point

Comment (CaselIT) [reply doc/build/orm/dataclasses.rst:41]: indeed. 

>  as of **March 9, 2023** is currently known to be fully supported

(btw today it's the 9th :)

**Comment by Viicos** [comment]
While I'm on this issue I encountered something weird while working with dataclasses: does column definition order matters when defining models? With the following one:

```python
class Model(Base):
    __tablename__ = "model"

    id: Mapped[int] = mapped_column(primary_key=True, init=False)
    uuid: Mapped[UUID] = mapped_column(default=None, insert_default=uuid4)
    name: Mapped[str] = mapped_column(
        String(255), doc="Name of the model", comment="Name of the model"
    )
    schemas: Mapped[List["Schema"]] = relationship(
        back_populates="model",
        doc="List of linked schemas",
    )
```

I get the following error: `TypeError: non-default argument 'name' follows default argument`

File: doc/build/orm/dataclasses.rst
Comment (Viicos) [reply doc/build/orm/dataclasses.rst]: https://github.com/sqlalchemy/sqlalchemy/pull/9463/commits/27ecc8321f0a6298b9bd7a56ac059c087d6247bc

Comment (Viicos) [reply doc/build/orm/dataclasses.rst]: > indeed.
> 
> > as of **March 9, 2023** is currently known to be fully supported
> 
> (btw today it's the 9th :)

I've added 6th as it was the release date of `mypy 1.1.1` ;), but I could change to today as well

Comment (CaselIT) [reply doc/build/orm/dataclasses.rst]: either one is fine. we could also use just the month. It's there just to give a general feel for time when the statement was written/updated

Comment (zzzeek) [reply doc/build/orm/dataclasses.rst]: we can let this poor submitter go free, it's fine

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 27ecc8321f0a6298b9bd7a56ac059c087d6247bc of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
New Gerrit review created for change 27ecc8321f0a6298b9bd7a56ac059c087d6247bc: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4503

**Comment by CaselIT** [comment]
> Edit: this behavior seems to come from the Python `dataclass` itself. Do you think this could be added in the docs for clarification?

I think it's implied by the docs, since all that logic is delegated to dataclasses. but I guess we could add a note here so that it gets mentioned https://github.com/sqlalchemy/sqlalchemy/issues/9410

**Comment by sqla-tester** [comment]
**Federico Caselli** (CaselIT) wrote:

depending on what https://github.com/sqlalchemy/sqlalchemy/issues/9467 turns out to be, mypy may not yet correctly support this pep



View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4503

**Review by zzzeek — CHANGES_REQUESTED** [review]
turns out we are wrong.  mypy has included pep-681 but they did not fix all issues with it, including the main issue that SQLAlchemy needs for it to work.

See https://github.com/python/mypy/issues/13856 , https://github.com/sqlalchemy/sqlalchemy/issues/9467 , https://github.com/python/mypy/issues/14868 .

So while this PR was to be celebratory, it instead needs to be a "if using mypy 1.1.1, you will need to install "type ignores" for all constructors, or stay on mypy 1.0.0 in the interim"  :(   or something else.  not really sure

**Comment by Viicos** [comment]
Nice catch, my mistake I only trusted the mypy's changelog, and did not apply mypy yet on a project I'm working on after switching to `MappedAsDataclass`. But I think it's fine to wait until a new fix release is issued before merging this one, hoping it won't take that long :)

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

we need to wait on this until mypy fixes https://github.com/python/mypy/issues/13856

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4503

**Comment by CaselIT** [comment]
I guess this could be closed at least unlit mypy fixes its implementation of pep 681

**Comment by CaselIT** [comment]
sorry @Viicos , but thanks for making us look to at mypy ultimately finding this issue.

hopefully we can redo the changes here soon

**Comment by sqla-tester** [comment]
**Federico Caselli** (CaselIT) wrote:

mypy has an issue with pep 681 and it needs to be fixed before merging this change

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4503

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4503 has been **abandoned**.  That means that at least for the moment I need to close this pull request. Sorry it didn't work out :(

**Pull request closed without merging**
